src/ann_solo/utils.py

```diff
@@ -1,104 +1,457 @@
-            formula=1, correction=0, full_output=True):
-        ssm.q = q
-        if q <= fdr:
-            yield ssm
+    logging.debug(
+        "Compute features for semi-supervised scoring from %d SSMs", len(ssms)
+    )
+    # Create a Dataset with SSM features.
+    features = pd.DataFrame(_compute_ssm_features(ssms))
+    # Remove invalid feature values (infinity).
+    for col in [
+        "mse_mz",
+        "mse_int",
+        "mse_mz_top5",
+        "mse_int_top5",
+        "manhattan",
+        "euclidean",
+        "chebyshev",
+        "canberra",
+    ]:
+        mask = np.isfinite(features[col])
+        features.loc[~mask, col] = features.loc[mask, col].max()
+    features["group"] = _get_ssm_groups(ssms, min_group_size) if grouped else 0
+    dataset = mokapot.dataset.LinearPsmDataset(
+        features,
+        target_column="is_target",
+        spectrum_columns="index",
+        peptide_column="sequence",
+        group_column="group",
+    )
+    if grouped:
+        logging.debug(
+            "Partitioned %d SSMs into %d groups",
+            len(ssms),
+            features["group"].nunique(),
+        )
+    # Define the mokapot model.
+    #   - Choice between a random forest, linear SVM, or no semi-supervised
+    #     learning.
+    #   - Features are preprocessed by standardizing them, removing
+    #     zero-variance features, and removing highly correlated features.
+    #   - We perform minimal tuning of the class weights and the random forest
+    #     tree depth.
+    if model is None:
+        logging.debug("Calculate q-values based on the cosine similarity")
+        # Calculate q-values based on the cosine similarity.
+        confidences = dataset.assign_confidence(features["cosine"], True)
+    else:
+        logging.debug(
+            "Train semi-supervised %s model and score SSMs", model.upper()
+        )
+        scaler = make_pipeline(
+            StandardScaler(),
+            VarianceThreshold(),
+            CorrelationThreshold(0.95),
+        )
+        if model == "svm":
+            clf = mokapot.model.PercolatorModel(
+                scaler, train_fdr=fdr, direction="cosine"
+            )
+        elif model == "rf":
+            clf = mokapot.Model(
+                GridSearchCV(
+                    RandomForestClassifier(random_state=1),
+                    param_grid={
+                        "max_depth": [3, 5, 7, 9, None],
+                        "class_weight": [
+                            None,
+                            {0: 0.1, 1: 1},
+                            {0: 0.1, 1: 10},
+                            {0: 1, 1: 0.1},
+                            {0: 1, 1: 10},
+                            {0: 10, 1: 0.1},
+                            {0: 10, 1: 1},
+                        ],
+                    },
+                    refit=False,
+                    cv=3,
+                    n_jobs=-1,
+                ),
+                scaler,
```

Comment:
**praise:** This is super cool! I hadn't actually thought about using a pipeline here.